tools/ci/check_grep2.py

```diff
@@ -0,0 +1,519 @@
+
+
+MULTIPLE_AREAS_RE = re.compile(r'"\w+" = \(\n[^)]*?/area/.+?,\n[^)]*?/area/.+?\)')
+
+
+def check_multiple_areas(content):
+    """
+    >>> content = '''"abc" = (
+    ... /turf/space,
+    ... /area/space)'''
+    >>> bool(check_multiple_areas(content))
+    False
+
+    >>> content = '''"abc" = (
+    ... /turf/space,
+    ... /area/foobar,
+    ... /area/space)'''
+    >>> bool(check_multiple_areas(content))
+    True
+    """
+    if match := MULTIPLE_AREAS_RE.search(content):
+        return f"At position {match.start(0)} multiple areas detected on the same tile. Please choose only one area."
+
+
+def check_common_spelling_mistakes(filename, lines):
+    for line in lines:
+        if "nanotransen" in line.lower():
+            return "ERROR: Misspelling of Nanotrasen detected in maps, please remove the extra N(s)."
+
+
+def check_515_proc_syntax(filename, lines):
+    if filename == IGNORE_515_PROC_MARKER:
+        return
+
+    for line in lines:
+        if re.match(r"\.proc/", line):
+            return "ERROR: Outdated proc reference use detected in code, please use proc reference helpers."
+
+
+def check_space_indentation(filename, lines):
+    """
+    Check specifically for space-significant indentation. Excludes dmdoc
+    block comment lines so long as there is an asterisk immediately after the
+    leading spaces.
+
+    >>> bool(check_space_indentation(None, ["  foo"]))
+    True
+
+    >>> bool(check_space_indentation(None, ["\\tfoo"]))
+    False
+    >>> bool(check_space_indentation(None, ["  * foo"]))
+    False
+    """
+    for line in lines:
+        if re.match(r"^ {2,}[^\*]", line):
+            return "ERROR: Space indentation detected, please use tab indentation."
+
+
+def check_mixed_indentation(filename, lines):
+    """
+    Check specifically for leading whitespace which contains a mix of tab and
+    space characters. Excludes dmdoc block comment lines so long as there is an
+    asterisk immediately after the leading whitespace.
+
+    >>> bool(check_mixed_indentation(None, ["\\t\\t foo"]))
+    True
+    >>> bool(check_mixed_indentation(None, ["\\t \\t foo"]))
+    True
+    >>> bool(check_mixed_indentation(None, ["\\t // foo"]))
+    True
+    >>> bool(check_mixed_indentation(None, [" \\tfoo"]))
+    True
+    >>> bool(check_mixed_indentation(None, ["  \\t  foo"]))
+    True
+
+    >>> bool(check_mixed_indentation(None, ["\\t  * foo"]))
+    False
+    >>> bool(check_mixed_indentation(None, ["\\t\\t* foo"]))
+    False
+    >>> bool(check_mixed_indentation(None, ["\\t \\t  * foo"]))
+    False
+    """
+    for line in lines:
+        if re.match(r"^(\t+ | +\t)\s*[^\s\*]", line):
+            return "ERROR: Mixed <tab><space> indentation detected, please stick to tab indentation."
+
+
+def check_trailing_newlines(filename, lines):
+    last_line = [x for x in lines][-1]
+    if not last_line.endswith("\n"):
+        return f"ERROR: File {filename} is missing a trailing newline"
```

Comment:
Can probably remove this and the filename parameters from the functions since it logs anyway

Reply:
The logging here is super cleaned up and much more consistent in terms of return values for each check, as well as output format, which is now always "filename:lineno: message".